### ISSUE TITLE
Version 4 Development and Changes

### ISSUE THREAD

**Issue by Balearica** [naptha/tesseract.js#662]
# Overview
While bug fixes continue to be released for Version 3, all breaking changes will be released in Version 4, which is currently under development in the branch named [dev/v4](https://github.com/naptha/tesseract.js/tree/dev/v4).  This branch should be usable at present by users eager to use any new features, however there is no guarantee that additional breaking changes will not be implemented.  Note that using this branch also requires using the Tesseract.js-core branch [dev/v4](https://github.com/naptha/tesseract.js-core/tree/dev/v4). 

# Summary
## Breaking Changes
1. `createWorker` is now async
   1. In most code this means `worker = Tesseract.createWorker()` should be replaced with `worker = await Tesseract.createWorker()`
   1. Calling with invalid `workerPath` or `corePath` now produces error/rejected promise (#654)
1. `worker.load` is no longer needed (`createWorker` now returns worker pre-loaded)
1. `getPDF` function replaced by `pdf` recognize option (#488)
   1. This allows PDFs to be created when using a scheduler
   3. See [browser ](https://github.com/naptha/tesseract.js/blob/dev/v4/examples/browser/download-pdf.html) and [node](https://github.com/naptha/tesseract.js/blob/dev/v4/examples/node/download-pdf.js) examples for usage

## Major New Features
1. Processed images created by Tesseract can be retrieved using `imageColor`, `imageGrey`, and `imageBinary`  options (#588)
   1. See [image-processing.html example](https://github.com/naptha/tesseract.js/blob/dev/v4/examples/browser/image-processing.html) for usage
1. Image rotation options `rotateAuto` and `rotateRadians` have been added, which significantly improve accuracy on certain documents
   1. See Issue #648 example of how auto-rotation improves accuracy
   5. See [image-processing.html example](https://github.com/naptha/tesseract.js/blob/dev/v4/examples/browser/image-processing.html) for usage of `rotateAuto` option
1. Tesseract parameters (usually set using `worker.setParameters`) can now be set for single jobs using `worker.recognize` options (#665)
   1. For example, a single job can be set to recognize only numbers using `worker.recognize(image, {tessedit_char_whitelist: "0123456789"})`
   2. As these settings are reverted after the job, this allows for using different parameters for specific jobs when working with schedulers
1. Initialization parameters (e.g. `load_system_dawg`, `load_number_dawg`, and `load_punc_dawg`) can now be set (#613)
   1. The third argument to `worker.initialize` now accepts either (1) an object with key/value pairs or (2) a string containing contents to write to a config file
   2. For example, both of these lines set `load_number_dawg` to 0:
      1. `worker.initialize('eng', "0", {load_number_dawg: "0"});`
      2. `worker.initialize('eng', "0", "load_number_dawg 0");`
 
## Other Changes
1. `loadLanguage` now resolves without error when language is loaded but writing to cache fails
   1. This allows for running in Firefox incognito mode using default settings (#609)
1. `detect` returns `null` values when OS detection fails rather than throwing error (#526)
2. Memory leak causing crashes fixed (#678)
3. Cache corruption should now be much less common (#666)

# Detail
## New Output Format Interface
A single, unified interface has been added for specifying all output formats.  `output` is now the 3rd argument to `recognize` (see example below).  This replaces the separate `getPDF` function, as well as various `setParameters` options (`tessjs_create_box`, `tessjs_create_hocr`, `tessjs_create_osd`, `tessjs_create_tsv`, and `tessjs_create_unlv`). 

```
const outputOpts = {
  text: true,
  blocks: true,
  hocr: true,
  tsv: true,
  box: false,
  unlv: false,
  osd: false,
  pdf: false,
  imageColor: false,
  imageGrey: false,
  imageBinary: false
};

const res = await worker.recognize(files[0], undefined, outputOpts);
```
Note: the default output formats (`text`, `blocks`, `hocr`, and `tsv`) are not changing between v3 and v4, so this change only impacts users who want non-default options.  This also means that users who want text *and* pdf outputs only need to specify `{pdf: true}`, as text is already a default.

**Comment by alaaeid1993** [comment]
is arabic avalible in load language

**Comment by Balearica** [comment]
@alaaeid1993 Yes the code for Arabic is `ara`.

**Comment by alaaeid1993** [comment]
> @alaaeid1993 Yes, the code for Arabic is `ara`. See [this file](https://github.com/naptha/tesseract.js/blob/master/src/constants/languages.js) for a list of languages and codes.

Yes, the Arabic letters are no problem, but the Arabic numbers do not appear correctly
im using angular version

**Comment by Balearica** [comment]
@alaaeid1993 This repo is for the JavaScript/webassembly port of Tesseract.  We do not make changes to the Tesseract OCR engine or language data (`.traineddata` files).  Issues related to accuracy are generally caused by Tesseract, and therefore outside of the scope of this repo.  

To confirm, you can install the Tesseract CLI (the main project) with an equivalent version (v5.3.0 as of this writing) and run with equivalent settings.  If you find that accuracy in Tesseract CLI is also unacceptable, then the issue is with Tesseract (not Tesseract.js), and you should look for a fix in the [main Tesseract repo](https://github.com/tesseract-ocr/tesseract).  If you find that Tesseract CLI produces correct results (with equivalent version/settings) but Tesseract.js does not, then we can discuss further here.

**Comment by andradatinta** [comment]
Could you please maybe provide the correct way of loading another language for the worker? I keep getting errors for the way I am using it right now because it's the only way I am seeing it being done in various tutorials. I am trying to work with the Romanian language and this is how I am currently doing it and it is not working: 
const worker = createWorker({
      logger: (m) => console.log(m), 
      langPath: "../tessdata", //for example
    });
    await worker.load();
    await worker.loadLanguage("ron");
    await worker.initialize("ron");
Thank you in advance, I am in desperate need of a solution for a personal project!

**Comment by Kumar6174** [comment]
![Screenshot from 2023-08-01 11-56-12](https://github.com/naptha/tesseract.js/assets/53547472/1bf2c770-7ef9-496f-a303-7f580ff503f5)

Tesseract.recognize(data);

i am passing data as blob data but don't know why showing me href error
[TypeError: Cannot read property 'href' of undefined]

**Comment by Balearica** [comment]
@Kumar6174 Please stop spamming this comment across unrelated Git Issues and delete all such comments you've made so far.  If there is an existing Git Issue that relates to your question, post it in that (single) Git Issue.  If no such Git Issue exists, then make a new one.
